### PR TITLE
feat(pricing): expand gateway max price to set per capability/model id

### PR DESCRIPTION
**Pull request by ad-astra-video**
This PR allows setting prices per capability and model id.

Current Implementation

Gateways currently use one MaxPrice set at launch with `maxPricePerUnit` flag and can be adjusted with the `setBroadcastConfig` endpoint of the cli webserver while the gateway is running. 

Enhancement Requested

Gateways on the AI subnet need the ability to filter orchestrators on price for different pipelines/model ids.  Transcoding would also benefit from this when different encoding codecs are enabled on the network.  The single MaxPrice needs to be updated to enable setting prices per capability (e.g. text-to-image and image-to-video) separately as well as allow the ability to price models within the capabilities separately.  The single MaxPrice will function as the overall default if no prices are set for specific pipelines and/or models.  The separate pricing needs to also allow setting a price per capability for all models in a capability.

Proposed implementation Changes

A `maxPricePerCapability` launch flag is added to accept json configuration of maximum prices for a capability/model.  The json configuration uses the `maxPricePerCapability` flag configuration to provide consistency in how prices are set on the network.  The `server.BroadcastCfg` is updated on the Gateway to store prices per capability/model id that can then be queried to provide the appropriate MaxPrice to use when selecting orchestrators in the selection algorithm.  The `server.BroadcastCfg` is also extended to add new functions for `GetCapabilitiesMaxPrice` and `GetCapabilityMaxPrice` to provide ability to get correct price with provided capabilities of the request. The `GetCapabilitiesMaxPrice` returns a total MaxPrice for the capabilities included in the request.  If no specific max prices is set for the capabilities requested, the MaxPrice is used.  `GetCapabilityMaxPrice` returns the max price set for that capability/model id or zero if not set.

The base `MaxPrice` and `SetMaxPrice` functions are updated to set pricing of H264 encoding capability to allow extending this to provide different prices for different codecs in the future.  ModelID is not used for transcoding so expected to always be "default".

### CODE REVIEW

File: cmd/livepeer/starter/starter_test.go
Comment (eliteprox) [inline cmd/livepeer/starter/starter_test.go]:
```suggestion
	assert.Equal(prices[0].Gateway, "default")
```
Update value for test

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:212]: Add return statements to prevent runtime error after failed request validation
```suggestion
				respond400(w, "pipeline and modelID must be set")
				return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:218]:
```suggestion
				respond400(w, "pipeline not supported")
				return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:226]:
```suggestion
					respond400(w, fmt.Sprintf("Error parsing pricePerUnit value: %s", maxPricePerUnit))
					return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:231]:
```suggestion
					respond400(w, fmt.Sprintf("Error parsing pixelsPerUnit value: %s", pixelsPerUnit))
					return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:235]:
```suggestion
					respond400(w, fmt.Sprintf("pixels per unit must be greater than 0, provided %v", pixelsPerUnit))
					return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:250]:
```suggestion
						respond400(w, errors.Wrap(err, "error converting price").Error())
						return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:260]:
```suggestion
				respond400(w, "maxPricePerUnit and pixelsPerUnit need to be set")
				return
```

File: server/handlers.go
Comment (eliteprox) [inline server/handlers.go:264]:
```suggestion
			respond400(w, "Node must be gateway node to set max price per capability")
			return
```

**Review by eliteprox — CHANGES_REQUESTED** [review]
Great work @ad-astra-video! I tested `-maxPricePerCapability` with AI model pricing and it works correctly. 

I also regression tested `-maxPricePerUnit` to ensure price controls still work with transcoding. Using the H264 capability as the default price index is a creative approach that should allow us to set prices per codec one day in the future. 

I tested `livepeer_cli` with both options 16, 15 and found the price changes also work correctly there. I made a few edits to `setMaxPriceForCapability` to prevent a runtime error and updated the test `TestMaxPricePerCapability`, see my suggestions below.

Overall, changes LGTM!

File: server/handlers.go
Comment (ad-astra-video) [reply server/handlers.go:212]: Agree

File: cmd/livepeer/starter/starter_test.go
Comment (ad-astra-video) [reply cmd/livepeer/starter/starter_test.go]: Agree

File: server/handlers.go
Comment (ad-astra-video) [reply server/handlers.go:250]: Agree

File: server/handlers.go
Comment (ad-astra-video) [reply server/handlers.go:235]: Agree

File: server/handlers.go
Comment (ad-astra-video) [reply server/handlers.go:260]: Agree

File: server/handlers.go
Comment (ad-astra-video) [reply server/handlers.go:264]: Agree

**Comment by ad-astra-video** [comment]
@eliteprox @rickstaa I agree with all suggestions and committed the suggested changes.

**Review by eliteprox — APPROVED** [review]
Changes LGTM. Approved pending review from @rickstaa

File: cmd/livepeer/starter/starter.go
Comment (victorges) [inline cmd/livepeer/starter/starter.go]: What is this field used for if, from what I understood, this is a config on the gateway itself? I thought this would make sense on an orchestrator config to set prices for specific gateways instead.

File: cmd/livepeer/starter/starter.go
Comment (victorges) [inline cmd/livepeer/starter/starter.go]: **nit:** Could be nice to support this defaulting for the `model_id` field as well. Then one can specify only the pipeline for it to be the default price for the pipeline

File: core/ai.go
Comment (victorges) [inline core/ai.go]: I wonder if this should be an explicit error instead (or the caller check this `Unused` value) to avoid silent bad configurations. Otherwise one might be missing a configuration that they swear they are making on the ir JSON config file which could have bad implications (e.g. paying more than they were willing to). In general I think it's better to fail explicitly for bad configs than to ignore them.

**Review by victorges — APPROVED** [review]

File: cmd/livepeer/starter/starter.go
Comment (ad-astra-video) [reply cmd/livepeer/starter/starter.go]: Removed Gateway field for now. There was a time that this logic was going to be re-used for the Orchestrator as well to set price per gateway/pipeline/model_id but I think we found a way to re-use the aiModels.json config parsing to set price per gateway at the Orchestrator. Can always add it back in if needed.  

Confirmed test still passed.

File: core/ai.go
Comment (ad-astra-video) [reply core/ai.go]: I agree that it would be better to update this to return `(Capability, error)` to make it more obvious it should be checked by the caller.

Updated and added a test for this function in ai_test.go

File: cmd/livepeer/starter/starter.go
Comment (ad-astra-video) [reply cmd/livepeer/starter/starter.go]: This is a good idea. I removed the p.Gateway check and changed it to setting p.ModelID to 'default' if not set specifically.

**Comment by ad-astra-video** [comment]
@victorges @rickstaa updates made for review comments.  Small merge conflict on a log line that looked the exact same resolved as well.

**Comment by rickstaa** [comment]
@ad-astra-video the code looks good 🚀. However I was unable to build this on my system 🤔. Could you check if you pushed the latest changes?

```bash
GO111MODULE=on CGO_ENABLED=1 CC="" CGO_CFLAGS="" CGO_LDFLAGS=" " go build -o "./" -tags "mainnet" -ldflags="-X github.com/livepeer/go-livepeer/core.LivepeerVersion=0.7.6-cdd74620" cmd/livepeer/*.go
# github.com/livepeer/go-livepeer/server
server/ai_session.go:206:3: undefined: Models
server/ai_session.go:208:29: capabilityConstraints[cap].Models undefined (type *"github.com/livepeer/go-livepeer/core".PerCapabilityConstraints has no field or method Models)
server/ai_session.go:209:14: undefined: core.NewCapabilitiesWithConstraints
server/broadcast.go:102:50: netCaps.CapabilityConstraints undefined (type *"github.com/livepeer/go-livepeer/net".Capabilities has no field or method CapabilityConstraints)
make: *** [Makefile:100: livepeer] Error 1
```

**Comment by ad-astra-video** [comment]
@rickstaa updated and re-tested this.

docker image build available at `docker pull adastravideo/go-livepeer:ai-video-set-max-price-per-capability-rebase`

File: cmd/livepeer/livepeer.go
Comment (rickstaa) [inline cmd/livepeer/livepeer.go]: Like the format! Do we want to add the `currency` field in there as well as an example?

Comment (rickstaa) [reply cmd/livepeer/livepeer.go]: Can we change this to snake case to keep the JSON convention. For this we can also change the `PricePerGateway` in this format (see https://linear.app/livepeer-ai/issue/AI-574/change-priceperbroadcaster-json-input-keys-to-recommended-conventions).

File: monitor/census.go
Comment (rickstaa) [inline monitor/census.go]: Can we name it `PricePerCapability`?

File: server/broadcast.go
Comment (rickstaa) [inline server/broadcast.go]: This one can be lowercase since it is not used outside this file :).

File: cmd/livepeer/livepeer.go
Comment (ad-astra-video) [reply cmd/livepeer/livepeer.go]: Updated. Will update pricePerGateway to same in separate PR.

File: monitor/census.go
Comment (ad-astra-video) [reply monitor/census.go]: Updated

File: server/broadcast.go
Comment (ad-astra-video) [reply server/broadcast.go]: Updated

File: server/broadcast.go
Comment (rickstaa) [inline server/broadcast.go]: As discussed in our chat this can be core.Capability_Unused.

Comment (ad-astra-video) [reply server/broadcast.go]: Updated

File: cmd/livepeer/starter/starter.go
Comment (rickstaa) [inline cmd/livepeer/starter/starter.go]: I think this field is a left over of another feature 🤔? Couldn't see it used.

**Review by rickstaa — APPROVED** [review]
LGTM thanks!

**Pull request merged**
